src/geyser/core.zig

```diff
@@ -0,0 +1,628 @@
+const std = @import("std");
+const builtin = @import("builtin");
+const sig = @import("../lib.zig");
+const bincode = sig.bincode;
+
+const c = @cImport({
+    // used to set/modify the pipe size
+    @cDefine("_GNU_SOURCE", {});
+    @cInclude("fcntl.h");
+
+    // used for mkfifo syscall
+    @cInclude("sys/types.h");
+    @cInclude("sys/stat.h");
+});
+
+const Account = sig.core.Account;
+const Pubkey = sig.core.Pubkey;
+const Slot = sig.core.Slot;
+const AccountWithoutData = sig.core.account.AccountWithoutData;
+const RecycleFBA = sig.utils.allocators.RecycleFBA;
+
+const PIPE_MAX_SIZE_PATH = "/proc/sys/fs/pipe-max-size";
+
+pub const AccountPayload = struct {
+    // used to know how much to allocate to read the full data slice
+    len: u64,
+    payload: VersionedAccountPayload,
+
+    const Self = @This();
+
+    pub fn deinit(self: *Self, allocator: std.mem.Allocator) void {
+        self.payload.deinit(allocator);
+    }
+};
+
+pub const VersionedAccountPayload = union(enum(u8)) {
```

Comment:
Why does this need to be versioned? Do you expect to have multiple versions going over the same stream?

Reply:
rn it doesnt need to be versioned but in the future when we roll out improvements / other payload types, this allows us to be backwards compatible

Reply:
What I mean is, when you do an upgrade, why is this approach to versioning necessary? I would only expect it is necessary when multiple versions need to coexist in memory simultaneously. Since geyser streams are ephemeral, not persistent, seems that you can just make the upgrade by switching to the new type and forgetting about the old one.

The advantage I see to versioning is that it presents a clear signal to readers when they have the wrong version and they need to upgrade. But that could be implemented with just an int field within a struct, doesn't need to be an enum.

Reply:
good point - yeah enums are not neccesary in this case and the same thing could be accomplished by ints - its just the simplest approach and what agave does rn

though the memory layout of enums will be wasteful since you need to alloc the largest struct, bincode serialization will basically be what you talk about (a single int to define which enum)

im not really hard stuck on either approach, but in terms of dev time required to add a new version, it would probably be easier to add a new enum